From f04190f9c0ba105fdaa68e69d843770bf182f8a5 Mon Sep 17 00:00:00 2001
From: Jianqun Xu <jay.xu@rock-chips.com>
Date: Mon, 28 Jun 2021 11:11:56 +0800
Subject: [PATCH 07/30] pinctrl: rockchip: Add rk3308b-s support

* From trm spec:
regval  RK3308B  RK3308BS
0       2mA      0mA
1       4mA      2mA
2       8mA      4mA
3      12mA      6mA
4                6mA
5                8mA
6               10mA
7               12mA

* But for the on board test:
regval  RK3308B  RK3308BS
0       5mA      0mA
1      10mA      5mA
2      20mA     10mA
3      25mA     15mA
4               15mA
5               20mA
6               25mA
7               30mA

So,
set RK3308B level0 to RK3308BS level1
set RK3308B level1 to RK3308BS level2
set RK3308B level2 to RK3308BS level5
set RK3308B level3 to RK3308BS level6

Change-Id: I50cf893ffa734dbd470932911a1bffdb4c9c41af
Signed-off-by: Jianqun Xu <jay.xu@rock-chips.com>
---
 drivers/pinctrl/pinctrl-rockchip.c | 52 ++++++++++++++++++++++++++++++++++++--
 1 file changed, 50 insertions(+), 2 deletions(-)

diff --git a/drivers/pinctrl/pinctrl-rockchip.c b/drivers/pinctrl/pinctrl-rockchip.c
index 67c8d9a..fe5dc0c 100644
--- a/drivers/pinctrl/pinctrl-rockchip.c
+++ b/drivers/pinctrl/pinctrl-rockchip.c
@@ -2684,6 +2684,29 @@ static int rockchip_get_drive_perpin(struct rockchip_pin_bank *bank,
 	return rockchip_perpin_drv_list[drv_type][data];
 }
 
+#define RK3308_SLEW_RATE_GRF_OFFSET		0x150
+#define RK3308_SLEW_RATE_BANK_STRIDE		16
+#define RK3308_SLEW_RATE_PINS_PER_GRF_REG	8
+
+static int rk3308_calc_slew_rate_reg_and_bit(struct rockchip_pin_bank *bank,
+					     int pin_num,
+					     struct regmap **regmap,
+					     int *reg, u8 *bit)
+{
+	struct rockchip_pinctrl *info = bank->drvdata;
+	int pins_per_reg;
+
+	*regmap = info->regmap_base;
+	*reg = RK3308_SLEW_RATE_GRF_OFFSET;
+	*reg += (bank->bank_num) * RK3308_SLEW_RATE_BANK_STRIDE;
+	pins_per_reg = RK3308_SLEW_RATE_PINS_PER_GRF_REG;
+
+	*reg += ((pin_num / pins_per_reg) * 4);
+	*bit = pin_num % pins_per_reg;
+
+	return 0;
+}
+
 static int rockchip_set_drive_perpin(struct rockchip_pin_bank *bank,
 				     int pin_num, int strength)
 {
@@ -2728,6 +2751,31 @@ static int rockchip_set_drive_perpin(struct rockchip_pin_bank *bank,
 		return ret;
 	}
 
+	if (soc_is_rk3308bs()) {
+		if (ret == 0)
+			ret = 1;
+		else if (ret == 1)
+			ret = 2;
+		else if (ret == 2)
+			ret = 5;
+		else if (ret == 3)
+			ret = 6;
+
+		data = 0x3 << (bit + 16);
+		rmask = data | (data >> 16);
+		data |= ((ret & 0x3) << bit);
+
+		ret = regmap_update_bits(regmap, reg, rmask, data);
+		if (ret)
+			return ret;
+
+		rk3308_calc_slew_rate_reg_and_bit(bank, pin_num, &regmap, &reg, &bit);
+		data = BIT(bit + 16) | (((ret > 3) ? 1 : 0) << bit);
+		rmask = BIT(bit + 16) | BIT(bit);
+
+		return regmap_update_bits(regmap, reg, rmask, data);
+	}
+
 	switch (drv_type) {
 	case DRV_TYPE_IO_1V8_3V0_AUTO:
 	case DRV_TYPE_IO_3V3_ONLY:
@@ -4130,7 +4178,7 @@ static int rk3308b_ctrl_data_re_init(struct rockchip_pin_ctrl *ctrl)
 	 * Special for rk3308b, where we need to replace the recalced
 	 * and routed arrays.
 	 */
-	if (soc_is_rk3308b()) {
+	if (soc_is_rk3308b() || soc_is_rk3308bs()) {
 		ctrl->iomux_recalced = rk3308b_mux_recalced_data;
 		ctrl->niomux_recalced = ARRAY_SIZE(rk3308b_mux_recalced_data);
 		ctrl->iomux_routes = rk3308b_mux_route_data;
@@ -4348,7 +4396,7 @@ static int rk3308b_soc_data_init(struct rockchip_pinctrl *info)
 	/*
 	 * Enable the special ctrl of selected sources.
 	 */
-	if (soc_is_rk3308b()) {
+	if (soc_is_rk3308b() || soc_is_rk3308bs()) {
 		ret = regmap_write(info->regmap_base, RK3308B_GRF_SOC_CON13,
 				   RK3308B_GRF_I2C3_IOFUNC_SRC_CTRL |
 				   RK3308B_GRF_GPIO2A3_SEL_SRC_CTRL |
-- 
2.7.4

